### ISSUE TITLE
Compile the training content loss: VGG runs twice a step and is never compiled

### ISSUE THREAD

**Issue by YeapJieShen** [YeapJieShen/single-image-super-resolution#227]
## The change

Let `torch.compile` cover the **training** content loss, not just the generator. Split out of
#226, which keeps the open-ended validation-cycle question.

## Why this half is separable

Compilation currently wraps one module:

```python
compiled = torch.compile(self.model, backend=backend, mode=...) if backend else None
```

and dispatch is gated on `self.training`. So the criterion is never compiled, in either phase.

The objection that gates the *validation* path does not apply here. Validation and benchmark
loaders present many image sizes, so a compiled module there risks recompiling per shape. **The
training path has exactly one shape**: the shipped adversarial config anchors the HR crop to the
discriminator's input size and its trainer block already records that every training step feeds
the same fixed crop. There is no recompilation bill to pay on this side.

## Where the work is

`VGG19FeatureLoss` runs the VGG stack **twice per step** — once on the prediction with gradients,
once on the target under `torch.no_grad()` — plus the backward through the first. The stack is
frozen (`.eval()`, `requires_grad_(False)`), so it is a fixed-shape, fixed-weight forward, which
is the friendliest possible compile target.

## Compile the terms, not the wrapper

`WeightedSumLoss` should stay eager, and the reason is that there is nothing in it to compile,
not that it is unsafe. Its `forward` is bookkeeping: a Python loop over a `ModuleDict`, two
runtime state probes (`is_inference()`, `is_inference_mode_enabled()`), a four-way branch
choosing `clone()` over `copy_()`, and dict writes. The arithmetic is one multiply and one add
per term. That branch is the shape that produces guard churn or a graph break, for no fusable
work.

It is already compile-safe by construction, and deliberately: the in-place `last_terms` write
exists because a replaying backend re-runs recorded kernels and never the Python line that binds
a name. That property should be preserved, not relied on to justify compiling the wrapper.

## The open question the profile decides

Compiling the generator and the loss as two separate units leaves a graph break exactly at the
generator-to-loss boundary, which is where fusion would pay most. A region covering both is more
promising and more invasive. **Do not choose without the measurement.**

## Config surface: a second flag, not a widened one

`compile_backend` today means "compile the generator". Widening its meaning in place would change
what an existing config does with no YAML edit — including the recorded run configs — so two runs
with identical YAML would stop being comparable. That is the exact failure this project's
provenance rules exist to prevent.

Add a separate opt-in instead, defaulting off, so what was compiled is legible from the config
that produced a number.

## Acceptance

- [ ] The content loss is compilable from a config, independently of the generator
- [ ] Default is off; an existing config's behaviour is unchanged
- [ ] `WeightedSumLoss` stays eager, and its per-term tags still track (the in-place write is
      what makes that true — a test should hold it)
- [ ] A before/after median ms/step on this machine, warmup excluded, on mains under the
      reference overlay

Blocked on the GPU until the running acceptance job finishes.

**Comment by YeapJieShen** [comment]
## Measured — and it overrules this issue's proposed shape

The issue says the region boundary must not be chosen without the measurement.
Here it is. Generator forward -> VGG54 content loss -> backward -> optimiser
step, at the shipped adversarial geometry (batch 16, 96x96 HR, 24x24 LR).
Medians, CUDA-synced, 30 warmup steps discarded so inductor's compilation is not
priced into the steady state. On mains, GPU otherwise idle.

| configuration | ms/step | vs eager |
|---|---:|---:|
| eager | 66.344 | 1.00x |
| generator compiled (what ships today) | 61.033 | 1.09x |
| **criterion compiled, generator eager** | 65.699 | **1.01x** |
| both, compiled as separate units | 64.195 | **1.03x** |
| **one region covering both** | **57.965** | **1.14x** |

### Three things this settles

**1. What this issue asks for does not pay.** "The content loss is compilable
from a config, independently of the generator" measures at **1.01x — 0.645 ms
on a 66 ms step**. The VGG stack does run twice per step, and it is still not
where compiling it independently buys anything.

**2. Compiling both as separate units is WORSE than compiling only the
generator** — 64.195 ms against 61.033. That is the graph break at the
generator-to-loss boundary, and it costs more than the criterion's own gain.
This issue predicted that break; it turns out to be net negative, not merely
suboptimal.

**3. The region wins, and it is the only configuration that beats what ships
today.** 1.14x, −8.4 ms.

### What I am not doing unattended, and why

Implementing the region means restructuring `_forward_sr` / `_step` — the method
whose own docstring calls it the single source of truth for the forward pipeline
and says not to re-implement it. A region returning only the loss also drops
`hr_cropped`, which that path produces. That is a seam change with a real
trade-off, so it wants a decision rather than a quiet commit.

Two further reasons to weigh it rather than take it:

- **The 1.14x is on a partial step.** A real adversarial step also runs the
  discriminator and the adversarial term, so the same 8.4 ms is a smaller share
  of the whole.
- **This GPU is SM-limited for inductor** — it logged *"Not enough SMs to use
  max_autotune_gemm mode"*. The ranking may not transfer to a larger card.

### Recommendation

**Close the "compile the criterion independently" framing as measured-and-declined**,
and keep the region as a separate, explicitly-scoped decision. If the region is
wanted, the config surface should still be a new opt-in defaulting off, exactly
as this issue argues — widening `compile_backend` in place would change what an
existing config does with no YAML edit.

**Comment by YeapJieShen** [comment]
## Closing: measured, and the proposed shape does not pay

Closing on the measurement above rather than on a code change. The region variant
is kept as an open design question outside the tracker; reopen this if it is
taken up.

### Why two separately-compiled units lost to compiling only the generator

`torch.compile` traces Python into an FX graph and hands it to inductor, which
fuses it into as few kernels as it can. Where the trace ends, the graph closes,
compiles, and control returns to eager Python. Compiling the generator and the
criterion as two separate calls creates such a seam **by construction** —
`processor.extract_target(hr)` and the argument passing between them run eager.

Four costs follow, and together they exceeded the gain:

1. **No fusion across the seam.** Inductor's largest win is folding pointwise
   work into a neighbouring kernel. The generator's tail (final convolution, then
   the processor's range mapping) and the VGG's head (normalisation, then its
   first convolution) are exactly that kind of work. Split, they cannot fuse.
2. **Forced materialisation.** The boundary activation must be written to global
   memory and read back rather than staying resident.
3. **Entry guards paid twice.** Every compiled region re-checks shapes, dtypes and
   `requires_grad` on entry. Two regions means two checks per step.
4. **The backward splits too.** Autograd records two compiled backward regions, so
   the same boundary cost is paid again in reverse.

Against that, the gain was small: a frozen VGG stack is dense convolution that is
already cuDNN-bound, with little pointwise work for inductor to fuse. Hence
1.01x for the criterion alone, and 1.03x for both split — **below the 1.09x that
compiling only the generator already achieves.**

Only a single region over both beats today (1.14x), because it is the one
arrangement with no seam at the generator-to-loss boundary — which is precisely
where the fusable pointwise work sits.

### Reopen trigger

Reopen if the region variant is taken up, or if these numbers are re-measured on
a GPU that is not SM-limited for inductor — this card logged *"Not enough SMs to
use max_autotune_gemm mode"*, so the ranking is not guaranteed to transfer.
